playwright/_impl/_browser_context.py

```diff
@@ -157,17 +157,23 @@ def _on_page(self, page: Page) -> None:
             page._opener.emit(Page.Events.Popup, page)
 
     def _on_route(self, route: Route, request: Request) -> None:
-        for handler_entry in self._routes:
-            if handler_entry.matches(request.url):
+        async def inner_route() -> None:
+            route_handlers = self._routes.copy()
+            for route_handler in route_handlers:
+                if not route_handler.matches(request.url):
+                    continue
+                if route_handler.will_expire:
+                    self._routes.remove(route_handler)
                 try:
-                    handler_entry.handle(route, request)
+                    handled = await route_handler.handle(route, request)
                 finally:
-                    if not handler_entry.is_active:
-                        self._routes.remove(handler_entry)
-                        if not len(self._routes) == 0:
-                            asyncio.create_task(self._disable_interception())
-                break
-        route._internal_continue()
+                    if len(self._routes) == 0:
+                        asyncio.create_task(self._disable_interception())
+                if handled:
+                    return
+            await route._internal_continue(is_internal=True)
+
+        asyncio.create_task(inner_route())
```

Comment:
I'd make on_route async and put create_task inside the lambda which calls on_route. Then page on_route can await browser_context._on_route and does not create nested tasks.

Reply:
Agreed! That's how it was in an earlier commit prior to debugging the greenlet issues, so let me go back to that commit.